sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/OrderedListUserState.java

```diff
@@ -0,0 +1,320 @@
+    // a VarLongCoder and the value is encoded with a LengthPrefixCoder.
+    // Refer to the comment in StateAppendRequest
+    // (org/apache/beam/model/fn_execution/v1/beam_fn_api.proto) for more detail.
+    private final KvCoder<Long, T> internalKvCoder;
+
+    public static <T> OrderedListUserState.TimestampedValueCoder<T> of(Coder<T> valueCoder) {
+      return new OrderedListUserState.TimestampedValueCoder<>(valueCoder);
+    }
+
+    @Override
+    public Object structuralValue(TimestampedValue<T> value) {
+      Object structuralValue = valueCoder.structuralValue(value.getValue());
+      return TimestampedValue.of(structuralValue, value.getTimestamp());
+    }
+
+    @SuppressWarnings("unchecked")
+    TimestampedValueCoder(Coder<T> valueCoder) {
+      this.valueCoder = checkNotNull(valueCoder);
+      this.internalKvCoder = KvCoder.of(VarLongCoder.of(), LengthPrefixCoder.of(valueCoder));
+    }
+
+    @Override
+    public void encode(TimestampedValue<T> windowedElem, OutputStream outStream)
+        throws IOException {
+      internalKvCoder.encode(
+          KV.of(windowedElem.getTimestamp().getMillis(), windowedElem.getValue()), outStream);
+    }
+
+    @Override
+    public TimestampedValue<T> decode(InputStream inStream) throws IOException {
+      KV<Long, T> kv = internalKvCoder.decode(inStream);
+      return TimestampedValue.of(kv.getValue(), Instant.ofEpochMilli(kv.getKey()));
+    }
+
+    @Override
+    public void verifyDeterministic() throws NonDeterministicException {
+      verifyDeterministic(
+          this, "TimestampedValueCoder requires a deterministic valueCoder", valueCoder);
+    }
+
+    @Override
+    public List<? extends Coder<?>> getCoderArguments() {
+      return Arrays.<Coder<?>>asList(valueCoder);
+    }
+
+    public Coder<T> getValueCoder() {
+      return valueCoder;
+    }
+
+    @Override
+    public TypeDescriptor<TimestampedValue<T>> getEncodedTypeDescriptor() {
+      return new TypeDescriptor<TimestampedValue<T>>() {}.where(
+          new TypeParameter<T>() {}, valueCoder.getEncodedTypeDescriptor());
+    }
+
+    @Override
+    public List<? extends Coder<?>> getComponents() {
+      return Collections.singletonList(valueCoder);
+    }
+  }
+
+  public OrderedListUserState(
+      Cache<?, ?> cache,
+      BeamFnStateClient beamFnStateClient,
+      String instructionId,
+      StateKey stateKey,
+      Coder<T> valueCoder) {
+    checkArgument(
+        stateKey.hasOrderedListUserState(),
+        "Expected OrderedListUserState StateKey but received %s.",
+        stateKey);
+    this.beamFnStateClient = beamFnStateClient;
+    this.timestampedValueCoder = TimestampedValueCoder.of(valueCoder);
+    this.request =
+        StateRequest.newBuilder().setInstructionId(instructionId).setStateKey(stateKey).build();
+  }
+
+  public void add(TimestampedValue<T> value) {
+    checkState(
+        !isClosed,
```

Comment:
When would the user see a closed state? Is it if they held onto it after the bundle closed or something like that?

Reply:
The way I implement close state in OrderedListUserState is consistent with the existing state.  The flag `isClosed` is set when the class method asyncClose() is called. 
https://github.com/apache/beam/blob/65aef1a3b48f36058cecf3479a17b4b129118e81/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/OrderedListUserState.java#L276

This class method is registered in stateFinalizers at `FnApiStateAccessor.java`
https://github.com/apache/beam/blob/65aef1a3b48f36058cecf3479a17b4b129118e81/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/FnApiStateAccessor.java#L929

It is then called after each processElement in `FnApiDoFnRunner.java`
https://github.com/apache/beam/blob/65aef1a3b48f36058cecf3479a17b4b129118e81/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java#L824-L842.